File: AspNetMvcUnleashed/0672329980_source/Chapter11/VB/MvcApplication1/MvcApplication1/Models/MovieRepository.vb

Public Class MovieRepository
    Inherits MovieRepositoryBase

    Private _entities As New MoviesDBEntities()
    Private _cache As Cache

    Public Sub New()
        _cache = HttpContext.Current.Cache
    End Sub


    Public Overrides Function ListMoviesCached() As IEnumerable(Of Movie)
        Dim movies = CType(_cache("movies"), IEnumerable(Of Movie))
        If movies Is Nothing Then
            movies = ListMovies()
            _cache("movies") = movies
        End If
        Return movies
    End Function


    Public Overrides Function ListMovies() As IEnumerable(Of Movie)
        Return _entities.MovieSet.ToList()
    End Function


    Public Overrides Sub CreateMovie(ByVal movieToCreate As Movie)
        _entities.AddToMovieSet(movieToCreate)
        _entities.SaveChanges()
        _cache.Remove("movies")
    End Sub
End Class

File: AspNetMvcUnleashed/0672329980_source/Chapter11/CS/MvcApplication1/MvcApplication1/Models/MovieRepository.cs
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Web.Caching;

namespace MvcApplication1.Models
{

    public class MovieRepository : MovieRepositoryBase
    {
        private MoviesDBEntities _entities = new MoviesDBEntities();
        private Cache _cache;

        public MovieRepository()
        {
            _cache = HttpContext.Current.Cache;
        }


        public override IEnumerable<Movie> ListMoviesCached()
        {
            var movies = (IEnumerable<Movie>)_cache["movies"];
            if (movies == null)
            {
                movies = ListMovies();
                _cache["movies"] = movies;
            }
            return movies;
        }


        public override IEnumerable<Movie> ListMovies()
        {
            return _entities.MovieSet.ToList();
        }


        public override void CreateMovie(Movie movieToCreate)
        {
            _entities.AddToMovieSet(movieToCreate);
            _entities.SaveChanges();
            _cache.Remove("movies");
        }
    }
}
